python-package/lightgbm/basic.py

```diff
@@ -0,0 +1,1080 @@
+        return preds, mat.shape[0]
+
+    def __pred_for_csr(self, csr, num_iteration, predict_type):
+        """
+        Predict for a csr data
+        """
+        nrow = len(csr.indptr) - 1
+        n_preds = self.num_class * nrow
+        if predict_type == C_API_PREDICT_LEAF_INDEX:
+            if num_iteration > 0:
+                n_preds *= num_iteration
+            else:
+                used_iteration = self.__num_total_model / self.num_class
+                n_preds *= used_iteration
+        preds = np.zeros(n_preds, dtype=np.float32)
+        out_num_preds = ctypes.c_int64(0)
+
+        ptr_indptr, type_ptr_indptr = c_int_array(csr.indptr)
+        ptr_data, type_ptr_data = c_float_array(csr.data)
+
+        _safe_call(_LIB.LGBM_BoosterPredictForCSR(
+            self.handle,
+            ptr_indptr, 
+            type_ptr_indptr,
+            csr.indices.ctypes.data_as(ctypes.POINTER(ctypes.c_int32)),
+            ptr_data,
+            type_ptr_data, 
+            len(csr.indptr), 
+            len(csr.data),
+            csr.shape[1], 
+            predict_type,
+            num_iteration,
+            ctypes.byref(out_num_preds),
+            preds.ctypes.data_as(ctypes.POINTER(ctypes.c_float))
+            ))
+        if n_preds != out_num_preds.value:
+            raise ValueError("incorrect number for predict result")
+        return preds, nrow
+
+
+class Dataset(object):
+    """Dataset used in LightGBM.
+
+    Dataset is a internal data structure that used by LightGBM
+    """
+
+    def __init__(self, data, label=None, max_bin=255, reference=None,
+        weight=None, group_id=None, predictor=None,
+        silent=False, params=None):
+        """
+        Dataset used in LightGBM.
+
+        Parameters
+        ----------
+        data : string/numpy array/scipy.sparse
+            Data source of Dataset.
+            When data is string type, it represents the path of txt file,
+        label : list or numpy 1-D array, optional
+            Label of the data
+        max_bin : int, required
+            max number of discrete bin for features 
+        reference : Other Dataset, optional
+            If this dataset validation, need to use training data as reference
+        weight : list or numpy 1-D array , optional
+            Weight for each instance.
+        group_id : list or numpy 1-D array , optional
+            group/query id for each instance. Note: if having group/query id, data should group by this id
+        silent : boolean, optional
+            Whether print messages during construction
+        params: dict, optional
+            other parameters
+        """
+        if data is None:
+            self.handle = None
+            return
+        self.data_has_header = False
+        """process for args"""
+        if params is None:
+            params = {}
+        self.max_bin = max_bin
+        self.predictor = predictor
+        params["max_bin"] = max_bin
+        if silent:
```

Comment:
I think *args and **kwargs is better to use outside of this class, like sklearn-liked interface